In [ ]:
---
layout: post
title: Binary Search
description: Team Teach Lesson HW
permalink: /binarysearch
comments: false
search_exclude: false
---

# Popcorn Hack #1

The answer is C, because in order for binary search to work as intended it needs to be looping through a sorted list. If it is not a sorted list, it wouldn't be able to find the correct answer because it can't "slice" the list into subsections.

# Popcorn Hack 2

The answer is B, because a sorted list is needed to successfully run a binary search. If it is not sorted, it will not be binary searched on.

# Popcorn Hack 3

In [ ]:
# Define a list of string elements
l = ["a", "b", "c", "d", "e", "f", "g"]

def binarySearch(lst, target):
    # This function performs a binary search on 'lst'
    # 'lst' is a list of sorted elements (we assume it's sorted in ascending order)
    # 'target' is the value we're looking for in 'lst'
    
    # Initialize the 'low' pointer to the start of the list (index 0)
    low = 0
    
    # Initialize the 'high' pointer to the end of the list (last index)
    high = len(lst) - 1
    
    # Continue searching as long as low <= high
    while low <= high:
        # Calculate the midpoint index between low and high
        # // ensures integer division
        mid = (low + high) // 2
        
        # Check if the element at the midpoint is our target
        if lst[mid] == target:
            # If it matches, return the midpoint index
            return mid
        elif lst[mid] < target:
            # If the midpoint value is less than the target,
            # move the 'low' pointer to one past the midpoint
            # to search the right half of the list
            low = mid + 1
        else:
            # Otherwise, if the midpoint value is greater than the target,
            # move the 'high' pointer to one less than the midpoint
            # to search the left half of the list
            high = mid - 1
    
    # If we exit the loop, the target isn't in the list
    return -1

# Call the binarySearch function and store the result
# We are searching for the string "f" in our list
result_index = binarySearch(l, "f")

# Print the result index to verify the function works as expected
print("Index of 'f':", result_index)


5

In [ ]:
import pandas as pd

# Load the dataset
data = pd.read_csv("school_supplies.csv")

# Drop rows with missing values
data_cleaned = data.dropna()

# Sort the data by 'Price'
data_sorted = data_cleaned.sort_values(by="Price")

# Extract sorted prices as a list
price_list = data_sorted["Price"].tolist()

# Preview the sorted data
print("First few rows of sorted data:")
print(data_sorted.head())
print("Original row count:", len(data))
print("Cleaned row count:", len(data_cleaned))

# Binary search function
def binary_search(sorted_list, target):
    left = 0
    right = len(sorted_list) - 1

    while left <= right:
        mid = (left + right) // 2
        if sorted_list[mid] == target:
            return mid  # Target found, return index
        elif sorted_list[mid] < target:
            left = mid + 1  # Search in the right half
        else:
            right = mid - 1  # Search in the left half

    return -1  # Target not found

# Prices to search for
prices_to_search = [1.25, 6.49, 10.00]

# Search for each price and print results
for price in prices_to_search:
    result = binary_search(price_list, price)
    if result != -1:
        print(f"Price {price} found at index {result}.")
    else:
        print(f"Price {price} not found.")


First few rows of sorted data:
        Product  Price
5        Eraser   0.50
14  Paper Clips   0.89
2        Pencil   0.99
9    Glue Stick   1.25
1           Pen   1.50
Original row count: 15
Cleaned row count: 15
Price 1.25 found at index 3.
Price 6.49 found at index 12.
Price 10.0 not found.


Below is a detailed explanation of how the updated code works:

1. **Data Preparation**  
   The dataset, represented as a pandas DataFrame (`data`), contains two columns: `Product` and `Price`. It includes 15 rows, each representing a school supply item and its corresponding price.  
   - The `data_cleaned` variable holds the same dataset after removing any rows with missing values. In this case, no rows were dropped since all entries are complete.  
   - The `data_sorted` variable contains the dataset sorted by the `Price` column in ascending order. This ensures that the binary search algorithm can operate correctly, as it requires a sorted list.

2. **Extracting Prices**  
   From the sorted DataFrame (`data_sorted`), the `Price` column is extracted and converted into a Python list (`price_list`). This list contains the prices in ascending order:  
   `[0.5, 0.89, 0.99, 1.25, 1.5, 1.75, 1.99, 2.25, 2.5, 2.99, 3.99, 4.99, 6.49, 15.75, 25.0]`.  
   This list serves as the input for the binary search function.

3. **Binary Search Algorithm**  
   The binary search function (`binary_search`) is implemented to efficiently locate a target value within a sorted list. Here's how it works:  
   - **Initialization**: Two pointers, `left` and `right`, are set to the start and end of the list, respectively.  
   - **Midpoint Calculation**: The midpoint index (`mid`) is calculated as the average of `left` and `right` (using integer division).  
   - **Comparison**: The value at the midpoint is compared to the target:  
     - If it matches the target, the function returns the index (`mid`).  
     - If the midpoint value is less than the target, the search range is narrowed to the right half by updating `left`.  
     - If the midpoint value is greater than the target, the search range is narrowed to the left half by updating `right`.  
   - **Termination**: The loop continues until `left` exceeds `right`. If the target is not found, the function returns `-1`.

4. **Testing the Function**  
   The code tests the binary search function with a list of target prices (`prices_to_search`): `[1.25, 6.49, 10.0]`.  
   - For each target price, the function is called with `price_list` and the target as inputs.  
   - The result is stored in the variable `result`. If the target is found, its index is printed. Otherwise, a message indicates that the target was not found.  
   - For example:  
     - `1.25` is found at index `3`.  
     - `6.49` is found at index `12`.  
     - `10.0` is not found, so the function returns `-1`.

5. **Output and Insights**  
   The output provides a clear demonstration of the binary search algorithm's efficiency and accuracy. It shows how the algorithm quickly narrows down the search range to locate the target values or determine their absence. This approach is significantly faster than a linear search, especially for larger datasets.

By combining data cleaning, sorting, and binary search, the code provides a robust workflow for efficiently searching through a dataset of school supplies and their prices.